### PR TITLE
Update PaddleScience APIs with 13 new basic modules

### PR DESCRIPTION
**Pull request by HydrogenSulfate**
### PR types
New features

### PR changes
APIs 

### Describe
1. Add new APIs in `ppsci/` directory, including
    1. `arch`
    2. `constraint`
    3. `data`
    4. `equation`
    5. `geometry`
    6. `autodiff`
    7. `loss`
    8. `metric`
    9. `optimizer`
    10. `solver`
    11. `utils`
    12. `validate`
    13. `visualize`
14. Change pre-commit hooks from mirror-yapf to isort + black

### CODE REVIEW

**Comment by cxxly** [comment]
按照PaddlePR规范修改标题和描述

File: paddlescience/data/__init__.py [hunk 1]
```diff
@@ -11,47 +11,44 @@
...
+from paddle.io import DataLoader
+from paddle.io import DistributedBatchSampler
 
-from .data_process import save_data, load_data
-from .trphysx_dataset import LorenzDataset, CylinderDataset, RosslerDataset
+from .data_process import load_data
+from .data_process import save_data
+from .trphysx_dataset import CylinderDataset
+from .trphysx_dataset import LorenzDataset
+from .trphysx_dataset import RosslerDataset
```
Comment (cxxly) [inline paddlescience/data/__init__.py:23]: pep8 https://peps.python.org/pep-0008/#imports

> Absolute imports are recommended, as they are usually more readable and tend to be better behaved (or at least give better error messages) if the import system is incorrectly configured (such as when a directory inside a package ends up on sys.path)

Comment (cxxly) [inline paddlescience/data/__init__.py:23]: image

File: paddlescience/data/__init__.py [hunk 2]
```diff
@@ -11,47 +11,44 @@
...
     """
-    Build the dataloader according to arguments 
-    dataset_name - name of the dataset 
-    batch_size - batch size, 
+    Build the dataloader according to arguments
+    dataset_name - name of the dataset
+    batch_size - batch size,
     shuffle - is shuffle data
-    drop_last -  """
+    drop_last -"""
```
Comment (cxxly) [inline paddlescience/data/__init__.py:35]: 文档格式符合任何规范

File: paddlescience/data/__init__.py [hunk 3]
```diff
@@ -11,47 +11,44 @@
...
 
-    assert dataset_name in [
-        'LorenzDataset', 'CylinderDataset', 'RosslerDataset'
-    ]
+    assert dataset_name in ["LorenzDataset", "CylinderDataset", "RosslerDataset"]
```
Comment (cxxly) [inline paddlescience/data/__init__.py:36]: 参数类型/值的检查使用异常，断言不抛出具体异常类型，仅用于一些测试、保证程序正确性的场景

File: ppsci/arch/base.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        for name, param in self.named_parameters():
+            if hasattr(param, "shape"):
+                num += np.prod(list(param.shape))
+            else:
+                logger.warning(f"{name} has no attribute 'shape'")
+        return num
+
+    def concat_to_tensor(
+        self, data_dict: Dict[str, paddle.Tensor], keys: List[str], axis=-1
+    ) -> List[paddle.Tensor]:
```
Comment (cxxly) [inline ppsci/arch/base.py:49]: 使用不可变类型，List -> Tuple

File: ppsci/constraint/base.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                f"name = {self.name}",
+                f"input_keys = {self.input_keys}",
+                f"output_keys = {self.output_keys}",
+                f"label_expr = {self.label_expr}",
+                f"label_dict = {self.label_dict}",
+                f"loss = {self.loss}",
+            ]
+        )
+        return _str
+
```
Comment (cxxly) [inline ppsci/constraint/base.py:44]: 直接return？

File: ppsci/constraint/base.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        data_dict = {}
+        for key in keys:
+            if key in alias_dict:
+                data_dict[alias_dict[key]] = np.asarray(
+                    raw_data_frame[key], "float32"
+                ).reshape([-1, 1])
+            else:
+                data_dict[key] = np.asarray(raw_data_frame[key], "float32").reshape(
+                    [-1, 1]
+                )
```
Comment (cxxly) [inline ppsci/constraint/base.py:60]: np.float32是否可能益处？

File: ppsci/constraint/base.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                f"label_expr = {self.label_expr}",
+                f"label_dict = {self.label_dict}",
+                f"loss = {self.loss}",
+            ]
+        )
+        return _str
+
+    def _load_csv_file(self, file_path, keys, alias_dict):
+        import pandas as pd
+
```
Comment (cxxly) [inline ppsci/constraint/base.py:47]: 如果pandas仅用于加载csv文件，python 内置csv是否更好，避免不必要的依赖

File: ppsci/constraint/interior_constraint.py [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+#                 if key in input_keys:
+#                     input[alias] = input.pop(key)
+#                 elif key in label_keys:
+#                     label[alias] = label.pop(key)
+#                 else:
+#                     raise ValueError(
+#                         f"key({key}) in alias_dict didn't appear "
+#                         f"in input_keys or value_keys"
+#                     )
+#         return input, label
```
Comment (cxxly) [inline ppsci/constraint/interior_constraint.py:274]: 这段注释是否要移除？

File: ppsci/constraint/supervised_constraint.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+        if not osp.exists(data_file):
+            raise FileNotFoundError(f"data_file({data_file}) not exist.")
+
+        self.input_keys = [
+            alias_dict[key] if key in alias_dict else key for key in input_keys
+        ]
+        self.output_keys = [
+            alias_dict[key] if key in alias_dict else key for key in label_keys
+        ]
+        if data_file.endswith(".csv"):
```
Comment (cxxly) [inline ppsci/constraint/supervised_constraint.py:63]: 是否有其它判断csv文件的方式？文件后缀不够通用，尤其对linux系统

File: ppsci/constraint/supervised_constraint.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        self,
+        data_file,
+        input_keys,
+        label_keys,
+        alias_dict,
+        dataloader_cfg,
+        loss,
+        weight_dict=None,
+        timestamps=None,
+        name="SupBC",
```
Comment (cxxly) [inline ppsci/constraint/supervised_constraint.py:52]: name能否用枚举类型封装？

**Comment by cxxly** [comment]
TODO： using type hints

File: ppsci/utils/__init__.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    "ExpressionSolver",
+    "AverageMeter",
+    "all_gather",
+    "convert_to_array",
+    "convert_to_dict",
+    "add_profiler_step",
+    "load_checkpoint",
+    "load_pretrain",
+    "save_checkpoint",
+]
```
Comment (cxxly) [inline ppsci/utils/__init__.py:42]: 此处API是否都面向用户公开？

File: ppsci/geometry/csg.py [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+
+    def is_inside(self, x):
+        return np.logical_or(self.geom1.is_inside(x), self.geom2.is_inside(x))
+
+    def on_boundary(self, x):
+        return np.logical_or(
+            np.logical_and(self.geom1.on_boundary(x), ~self.geom2.is_inside(x)),
+            np.logical_and(self.geom2.on_boundary(x), ~self.geom1.is_inside(x)),
+        )
+
```
Comment (cxxly) [inline ppsci/geometry/csg.py:52]: geometry相关数据结构及运算能否用Paddle

File: ppsci/gradient/__init__.py [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+from ppsci.gradient.grad import clear
+from ppsci.gradient.grad import hessian
+from ppsci.gradient.grad import jacobian
```
Comment (cxxly) [inline ppsci/gradient/__init__.py:18]: `autodiff/ad` 或许是更好的包名，严格来讲，只有vector-scalar型函数才叫gradient

File: ppsci/gradient/grad.py [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+        return self.Hs[key](i, j)
+
+    def clear(self):
+        """Clear cached Hessians."""
+        self.Hs = {}
+
+
+jacobian = Jacobians()
+hessian = Hessians()
+
```
Comment (cxxly) [inline ppsci/gradient/grad.py:145]: 根据设计文档中结论，这两个API最终需要下沉到Paddle，前期可以在PaddleScience验证

File: ppsci/loss/l1.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+                loss *= weight_dict[key]
+            if "area" in output_dict:
+                loss *= output_dict["area"]
+
+            if self.reduction == "sum":
+                loss = loss.sum()
+            elif self.reduction == "mean":
+                loss = loss.mean()
+            losses += loss
+        return losses
```
Comment (cxxly) [inline ppsci/loss/l1.py:44]: 此处不使用paddle.nn.L1Loss原因是？

File: ppsci/loss/mse.py [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+                loss *= weight_dict[key]
+            if "area" in output_dict:
+                loss *= output_dict["area"]
+
+            if self.reduction == "sum":
+                loss = loss.sum()
+            elif self.reduction == "mean":
+                loss = loss.mean()
+            losses += loss
+        return losses
```
Comment (cxxly) [inline ppsci/loss/mse.py:44]: 此处不适用paddle.nn.MSELoss原因是？

File: ppsci/solver/printer.py [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+#     # compute metric
+#     if trainer.eval_metric_func is not None:
+#         metric_dict = trainer.eval_metric_func(out, batch[-1])
+#         for key in metric_dict:
+#             if key not in trainer.eval_output_info:
+#                 trainer.eval_output_info[key] = misc.AverageMeter(key, "7.5f")
+#             trainer.eval_output_info[key].update(
+#                 float(metric_dict[key]), batch_size
+#             )
+
```
Comment (cxxly) [inline ppsci/solver/printer.py:62]: 注释是否需要

File: ppsci/solver/solver.py [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+
+    Args:
+        cfg (AttrDict): Configuration parsed from yaml.
+        mode (str, optional): Running mode. Defaults to "train".
+    """
+
+    def __init__(self):
+        pass
+
+    def initialize_from_config(self, cfg, log_level, mode="train"):
```
Comment (cxxly) [inline ppsci/solver/solver.py:46]: 是否有其他构造方式，如果没有，用默认init，cfg作为参数

File: ppsci/arch/base.py (hunk 4 above)
Comment (HydrogenSulfate) [reply ppsci/arch/base.py:49]: Done

Comment (HydrogenSulfate) [reply ppsci/arch/base.py]: Done

File: ppsci/constraint/base.py (hunk 5 above)
Comment (HydrogenSulfate) [reply ppsci/constraint/base.py]: Done

File: ppsci/constraint/base.py (hunk 7 above)
Comment (HydrogenSulfate) [reply ppsci/constraint/base.py]: Done

File: ppsci/constraint/interior_constraint.py (hunk 8 above)
Comment (HydrogenSulfate) [reply ppsci/constraint/interior_constraint.py]: Deleted

File: ppsci/constraint/supervised_constraint.py (hunk 9 above)
Comment (HydrogenSulfate) [reply ppsci/constraint/supervised_constraint.py:60]: try cache

File: ppsci/loss/l1.py (hunk 15 above)
Comment (HydrogenSulfate) [reply ppsci/loss/l1.py:44]: weight_dict会与loss逐点相乘，需要把error计算和reduce分开

File: ppsci/loss/mse.py (hunk 16 above)
Comment (HydrogenSulfate) [reply ppsci/loss/mse.py:44]: 同L1Loss

File: ppsci/solver/printer.py (hunk 17 above)
Comment (HydrogenSulfate) [reply ppsci/solver/printer.py]: Deleted

File: ppsci/gradient/__init__.py (hunk 13 above)
Comment (HydrogenSulfate) [reply ppsci/gradient/__init__.py]: rename `gradient/grad.py` to `autodiff/ad.py`

File: ppsci/utils/__init__.py (hunk 11 above)
Comment (HydrogenSulfate) [reply ppsci/utils/__init__.py:32]: 已删减

**Review by cxxly — APPROVED** [review]
LGTM with todo list:

- 【P0】refactor solver construct
- 【P1】add unittest
- 【P1】replace geometry with paddle API
- 【P2】support type hints
